src/librustc_mir/hair/mod.rs

```diff
@@ -28,6 +26,8 @@ use self::cx::Cx;
 
 pub mod cx;
 
+pub use rustc_const_eval::pattern::{BindingMode, Pattern, PatternKind, FieldPattern};
```

Comment:
Is moving parts of HAIR to `rustc_const_eval` a good idea? What about doing the opposite?
That is, move match checking to `rustc_mir` and perhaps even integrate it with MIR match building.